### ISSUE TITLE
JAWS 2019 does not read aria-required for checkboxes as expected

### ISSUE THREAD

**Issue by mberrisford** [FreedomScientific/VFO-standards-support#185]
### Summary

> Brief description of the issue. Please list any specific steps.
1. Go to any form input checkbox with the aria-required="true" attribute. 
`<input type="checkbox" aria-required="true">`
2. Tab or Shift+Tab to  the checkbox.

### Expected result

> What did you expect to happen?
"Required" is read out lout after label when you tab to the checkbox.

### Actual result

> What actually did happen?
"Required" is not read out loud unless you interact with the checkbox using the space bar to check or uncheck it. 

#### JAWS version and build number
2019.1903.47 and previous version.

#### Operating System and version
Win10

#### Browser and version:
Tested on IE11, FF, Chrome.

**Comment by JasonAment** [comment]
I am unable to reproduce this in the latest version of JAWS.  2019.1903.47 
`<form>
      <label for="test1">test case 1</label>
      <input id="test1" type="checkbox" aria-required="true">
</form>
<label for="test2">test case 2</label>
<input id="test2" type="checkbox" aria-required="true">`

**Comment by mberrisford** [comment]
Have you tried it in Chrome and FF? I was actually wrong about IE. It does work for me in IE but not Chrome or FF. Sorry about the confusion.

**Comment by JAWS-test** [comment]
The error description is correct (no output of aria-required at checkboxes at Tab Navigation in Firefox and Chrome). However, this is no JAWS bug, but the ARIA specification does not allow aria-required for checkboxes: http://www.w3.org/TR/wai-aria-1.1/#aria-required. The same applies to the HTML specification with the required attribute. Required can only be form fields that actually expect a user input, this is not the case with checkboxes (only the status can be changed here). In this respect, aria-required should not be output at all and the error would rather be the output during operation mit space bar and linear reading (and with Tab in IE 11). 

The ticket should be closed or modified: aria-required and required should not be output for checkboxes and other elements for which it is not intended.

**Comment by scottaohara** [comment]
@mberrisford do you find the same issue with the required state not being announce per [this required checkbox test page](https://scottaohara.github.io/tests/html-inputs/input-checkbox-required.html)?  

In my testing all tests did produce a "required" announcement with the latest version of JAWS 2019, released in June, with default settings.

@JAWS-test regarding:
>Required can only be form fields that actually expect a user input, this is not the case with checkboxes

This is not true.  `required` can be used on checkbox inputs in HTML.

Also of note, ARIA 1.2 is being updated (see [latest working draft](https://w3c.github.io/aria/#checkbox)) to have parity with HTML, and `aria-required` will be an allowed attribute for `role=checkbox`.

**Comment by mberrisford** [comment]
Thanks for that test page @scottaohara 

I am no longer able to reproduce this issue with the latest version of JAWS 2019.1906.10 with any browser. (I was able to with the previous version of JAWS in Chrome and FF only.)

@JAWS-test ...
"required" is widely used for radio buttons as well as checkboxes. Can you point me to a resource that backs up your claim that it's invalid HTML? 

I do see that aria-required for checkboxes is omitted from the current ARIA specs.

Thanks for taking the time.

Martin

**Comment by JAWS-test** [comment]
@mberrisford In the current version of the ARIA specification (1.1), aria-required is not possible with checkboxes. This also applied to HTML checkboxes required for HTML 5.0. Unfortunately, I have not checked that this has changed with HTML 5.2 and will change with ARIA 1.2. Sorry for my mistake. However, I am of the opinion that the old specification was better than the new/future one, since required is not useful for a checkbox. A checkbox can only have 2 states. If it is required, it can mean that the user has to think about whether to click it or not (but that's always the case), or that he or she has to click it (but then you don't need a checkbox either).

However, I also know that in practice the required attribute for checkboxes is needed for these "I accept the terms and conditions" queries. That could be implemented technically also in a different way, but it has become established...

**Comment by scottaohara** [comment]
[HTML 5.0](https://www.w3.org/TR/html50/forms.html#checkbox-state-(type=checkbox)) allowed for the required attribute on checkboxes. I assume to allow for the situation you mentioned, as that would be a logical use case for it.

Conversations about the usefulness of it aside, it's allowed and well supported, as the test case demonstrates.

**Comment by JAWS-test** [comment]
@scottaohara: Regarding required and HTML: I have now re-examined the old specifications and found that I had the wrong memory, either confused with readonly (which is not allowed with checkboxes) or misled by the assumption that ARIA and HTML largely match.

Whether required well-supported is, one can still argue about it. With JAWS I notice the following problems with required:
* IE 11: No output when reading linearly
* IE 11: No output of the error message after submitting the form
* Chrome, IE 11 etc.: Output also for button and links (where I don't think it's really allowed and useful)
* Chrome: inconsistent output if required at fieldset: for linear reading e.g. not for input fields, but for checkboxes
* and the incorrect output of aria-required described in this ticket, which was not fixed until the last release of JAWS
